runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/CoderTranslator.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction;
+
+import java.util.List;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.coders.Coder;
+
+/**
+ * An interface that translates coders to components and back.
+ *
+ * <p>This interface is highly experimental, and incomplete. Coders must in the general case have
+ * the capability to encode an additional payload, which is not currently supported. This exists as
+ * a temporary measure.
+ */
+@Experimental(Kind.CORE_RUNNERS_ONLY)
+public interface CoderTranslator<T extends Coder<?>> {
```

Comment:
Just to make sure I have this in the right frame in my head - we only need `Any` versus `bytes` in the event we need to do some dynamic typing on the field (basically, checking what type it is and "downcasting" appropriately).

`Struct` seems a better fit than `Any`, or even a less general `map<string, Any>`/`map<string, bytes>` (the proto can have a `Struct` even if Java uses a simple `Map`). I believe every interesting case will be a dict of named params that may or may not be complex objects.

In that case we could reasonably have the `CoderTranslator` return some JSON-like object (or map) so it is OK to have in the SDK core and in runners-core-construction we convert to proto. Perhaps `CoderParameters getNamedParameters(CoderT coder)` where `CoderParameters` is a class of our own devising, just to decouple from `Map` and `GenericJson`, etc.

p.s. I also agree with @tgroh's point that this unblocks thinning the SDK which is most important for API freeze. We just need to make sure not to introduce costly dependencies on temporary works.